src/internal/header.rs

```diff
@@ -1,20 +1,66 @@
 //! RPM package headers
 
-use super::{tag::Tag, td::TagData};
-use crate::Package;
+use super::archive::{RpmReturnCode, RpmErrorKind};
+use super::{tag::Tag, td::TagData, ts::TransactionSet};
+use std::ffi::CString;
 use std::mem;
+use std::os::unix::prelude::OsStrExt;
+use std::path::Path;
 
 /// RPM package header
 pub(crate) struct Header(*mut librpm_sys::headerToken_s);
 
 impl Header {
+    pub(crate) fn new() -> Self {
+        let ffi_header = unsafe { librpm_sys::headerNew() };
+        assert!(!ffi_header.is_null());
+
+        // No need to increment refcount, starts with refcount=1
+        Header(ffi_header)
+    }
+
+    /// Create a Header handle in Rust from a raw pointer
+    ///
+    /// SAFETY: The input pointer must not be used after passing ownership from Rust, except for dropping
+    /// the live reference if one existed.
     pub(crate) unsafe fn from_ptr(ffi_header: librpm_sys::Header) -> Self {
         assert!(!ffi_header.is_null());
         // Increment librpm's internal reference count for this header
         librpm_sys::headerLink(ffi_header);
         Header(ffi_header)
     }
 
+    pub(crate) fn from_file(path: &Path) -> Result<Self, RpmErrorKind> {
+        let filename = CString::new(path.as_os_str().as_bytes()).unwrap();
+        let fmode = CString::new("r.ufdio").unwrap();
+
+        let fd: librpm_sys::FD_t = unsafe { librpm_sys::Fopen(filename.as_ptr(), fmode.as_ptr()) };
+        // GlobalTS.lock() ?  Or can the transaction sets be independent because we're not touching a global database?
+        let ts = TransactionSet::create();
```

Comment:
@tarcieri This PR isn't ready for review (it doesn't even compile) but I have a question about how this works.

Reply:
Yeah, I should probably write up some documentation on how those work, at least in the current form.

It's one of those things that went through multiple rewrites trying different strategies after long, protracted discussions with RPM maintainers, and even then there are still bugs in the current implementation I believe.

Reply:
To perhaps more directly answer your question, the current locking model uses `GlobalTS` for everything, in all cases.